[FEAT] 상담 기능 API 개발

yjy323 opened the pull request:
## 수정 사항
> 2025.11.15
기존 코드를 너무 급하게 짜서 품질이 엉망이더라구요 ㅜㅜ 죄송합니다.
이번에 수정을 하면서는 DTO 객체에 toEntity 메서드, from static 메서드를 두어서 응집도를 높이고 Service Layer에 Mapper 메서드를 모두 지웠습니다.

## 주요 변경 사항
- ConsultationRequest → ConsultationReservation으로 도메인명 변경
- 모든 DTO 클래스명을 Request/Response 패턴으로 통일
  - CreateConsultationDto → CreateConsultationRequest
  - UpdateConsultationNoteDto → UpdateConsultationNoteRequest
  - ConsultationRequestResponse → ConsultationReservationResponse
- 상담 예약 승인/거절 API를 별도 엔드포인트로 분리
  - /consuntations/reservations/{id}/approve, /consuntations/reservations/{id}/reject
- @Pattern 어노테이션으로 전화번호 형식 검증 추가
- @JsonFormat으로 날짜/시간 형식 표준화
- DTO 변환 로직을 각 클래스 내부로 캡슐화
- 상담 예약, 상담 Status 변경 로직 단순화

추가로 ERD와 API는 오늘 저녁까지 수정해두겠습니다! 이어서 사건, 수임 API 순으로 리뷰할게요.

## 🪺 개요 
상담 예약 - 상담 변호사 배정 - 상담 진행(의뢰인 개인정보 작성 + 상담 노트 + 파일 업로드) - 상담 완료 - 사건 생성까지의 프로세스를 처리할 수 있는 API를 개발했습니다.

## 💻 작업 사항 
  상담 도메인 핵심 기능 구현
  - 상담 예약 관리: ConsultationRequest 엔티티 및 상태 관리 (PENDING → APPROVED/REJECTED)
  - 상담 진행 관리: Consultation 엔티티 및 상태 전이 (CONSULTATION_WAITING → CONSULTATION_IN_PROGRESS → COMPLETED)  
  - 상담 노트: ConsultationNote 엔티티로 상담 내용 체계적 기록
  - 파일 관리: ConsultationFile과 FileEntity를 통한 상담 관련 서류 업로드/관리
  - 고객 정보: Client 엔티티 개선 및 Gender enum 타입 적용
  - 상담-사건 연동: 상담 완료 시 자동 사건(Cases) 생성

## 🌱 Issue Number
- #10 

## ✅ 셀프 체크리스트
- [x] PR 제목을 형식에 맞게 작성했나요 ?
- [x] 브랜치 전략에 맞는 브랜치에 PR을 올리고 있나요 ?
- [x] 테스트는 잘 통과했나요 ?
- [x] 빌드에 성공했나요 ?
- [x] 본인을 Assign 해주세요.
- [x] 리뷰할 팀원들을 Request 해주세요.
- [x] 해당 PR에 맞는 label을 붙여주세요.

## 🙏 To Reviewers

yjy323 committed:
refactor: Client.Gender를 String에서 enum 타입으로 리팩토링

yjy323 committed:
Merge branch 'develop' of https://github.com/Hae-il/BE into feat/consultation-api-#10

yjy323 committed:
feat: 상담 예약 관리 API 구현 #10

- 상담 예약 신청 API 추가
- 상담 예약 목록 조회 API 구현
- 상담 예약 상세 조회 API 구현
- 상담 예약 상태 변경 API 구현 (승인/거절)
- ConsultationRequest 도메인 로직 추가
- ConsultationService 비즈니스 로직 구현
- 예외 처리 및 에러 코드 정의
- DTO 클래스 및 Repository 추가

yjy323 committed:
feat: 상담 관리 API 구현 #10

- 상담 시작 API 추가 (승인된 예약 기반)
- 상담 목록/상세 조회 API 구현
- 상담 상태 변경 API 구현
- Consultation 도메인 및 비즈니스 로직 추가
- Client 도메인 확장 및 자동 생성 기능
- 상담 관련 DTO 클래스 추가

yjy323 committed:
feat: 상담 노트 관리 API 구현 #10

- 상담 노트 작성 API 추가
- 상담 노트 조회 API 구현 (상담별 노트 목록)
- 상담 노트 수정 API 구현 (작성자 권한 체크)
- ConsultationNote 도메인 및 비즈니스 로직 추가
- 상담 진행 상태별 노트 작성 권한 관리
- 상담 노트 관련 DTO 및 Repository 추가

yjy323 committed:
feat: 상담 파일 업로드 및 관리 기능 구현 #10

- 상담 파일 업로드 API 엔드포인트 추가
- 상담 파일 목록 조회 API 엔드포인트 추가
- ConsultationFile 도메인 및 레포지토리 구현
- 파일 서비스 통합으로 파일 관리 기능 제공

yjy323 committed:
feat: 상담 기능 추가 구현 완료 #10

- Cases 엔티티에 consultation 연관관계 추가
- CasesService에 상담 관련 조회 메서드 추가
- Consultation 엔티티 구조 개선
- ConsultationStatus enum 타입 추가
- ConsultationService에 CRUD 기능 완전 구현

yjy323 committed:
Merge branch 'develop' of https://github.com/Hae-il/BE into feat/consultation-api-#10

yjy323 committed:
fix: Cases conflict 해결

yjy323 committed:
feat: Cases와 Consultation 엔티티 관계 개선 #10

- Cases 엔티티의 consultationId를 Consultation 객체 참조로 변경
- 오타 수정: occuredDate -> occurredDate
- User 엔티티의 매핑 관계 정리 및 final 키워드 추가

yjy323 committed:
feat: Cases와 Consultation 엔티티 관계 개선 #10

- Cases 엔티티의 consultationId를 Consultation 객체 참조로 변경
- 오타 수정: occuredDate -> occurredDate
- User 엔티티의 매핑 관계 정리 및 final 키워드 추가

yjy323 commented on src/main/java/com/haeil/be/cases/domain/Cases.java:
지금은 사건에서 Client를 참조하려면 Consultation.Client로 참조해야 하는데, 필드를 추가해서 Client를 Consultation으로 부터 주입받는 것이 좋을 것 같습니다.

yjy323 committed:
Merge branch 'feat/consultation-api-#10' of https://github.com/Hae-il/BE into feat/consultation-api-#10

yjy323 reviewed (commented):
1차적으로 검토했습니다!
Controller, Service가 좀 비대한 것 같은데 리팩토링은 추후에 고려하겠습니다.

Notion에 상담 시나리오랑 API 명세 이어서 업데이트 하겠습니다.

erika0915 commented:
수정된 erd를 혹시 erdcloud에 반영해주실 수 있나요 ? !! 

> 1차적으로 검토했습니다! Controller, Service가 좀 비대한 것 같은데 리팩토링은 추후에 고려하겠습니다.
> 
> Notion에 상담 시나리오랑 API 명세 이어서 업데이트 하겠습니다.

erika0915 commented on src/main/java/com/haeil/be/client/domain/Client.java:
```diff
@@ -36,12 +38,56 @@ public class Client extends BaseEntity {
+        this.jobTitle = jobTitle;
+    }
+
+    public static Client fromRequest(
```
요거는 dto (또는 service 부분)에 들어가도 좋을 것 같아요 ! 엔티티 안에는 순수하게 도메인 로직에만 집중하고, 이 부분은 dto나 service에서 처리하는게 어떨까요 ? 저는 주로 dto에서 관리하는 것 같긴합니당 !

yjy323 commented:
> 수정된 erd를 혹시 erdcloud에 반영해주실 수 있나요 ? !!
> 
> > 1차적으로 검토했습니다! Controller, Service가 좀 비대한 것 같은데 리팩토링은 추후에 고려하겠습니다.
> > Notion에 상담 시나리오랑 API 명세 이어서 업데이트 하겠습니다.

네네! 반영하겠습니다!

erika0915 commented on src/main/java/com/haeil/be/file/domain/FileEntity.java:
```diff
@@ -0,0 +1,47 @@
+@Entity
+@Getter
+@NoArgsConstructor
+public class FileEntity extends BaseEntity {
```
우리 서비스에 업로드 되는 모든 파일을 관리하는 엔티티가 되는거죠 ? 좋은 것 같아요 ! 사건에서도 파일들 관리하는 부분이 있으니깐 범용성이 좋아 보입니다 ~! 👍🏻

erika0915 commented on src/main/java/com/haeil/be/consultation/controller/ConsultationController.java:
```diff
@@ -1,12 +1,153 @@
+
+    private final ConsultationService consultationService;
+
+    @Operation(summary = "상담 예약 신청", description = "새로운 상담 예약을 신청합니다.")
```
요거는 임의로 구현한 api라고 이해하면 되죵 ?

yjy323 replied:
네넵 이제 API 명세에 반영하겠습니다.

erika0915 reviewed (approved):
전반적으로 확인했습니다 !! 그 하나 궁금한거는 어떤 api는 {id} 이고 어떤 api는 {consultationId} 더라고요 ! 혹시 의도하신건지 궁금합니당 !! 아니라면, 하나로 동일해도 좋을 것 같다는 생각도 있습니당 !! 프론트에서 연결할 때 헷갈릴 수 있을 것 같아요 ! 그리고 api 명세서 안에 눌러서 안에 조금 세부적으로 작성해주시면 api 잘 돌아가는지 확인할 때 더 편할 것 같아요 ~ api 돌려보다가 요청 파라미터 값들이 어떤 형태여야 하는지 명확하지 못해서 좀 고생을 했어용 ㅠㅠ  많은 api 개발하시느라 고생하셨어요 👍🏻🥇💯

erika0915 commented on src/main/java/com/haeil/be/consultation/dto/request/CreateConsultationDto.java:
```diff
@@ -0,0 +1,36 @@
+    @NotNull(message = "의뢰인 정보는 필수 입력 항목입니다.")
+    private ClientInfoDto clientInfo;
+
+    private LocalDateTime consultationDate;
```
<img width="497" height="328" alt="Image" src="https://github.com/user-attachments/assets/efff3ef8-1c1f-4728-9bd3-cc883aa6f7e0" />

여기 보니간 LocalDateTime로 요청에서 파라미터로 받는 부분에서 (여기 뿐만 아니라 다른 부분도 있더라고요 !) 요청 형태를 이런식으로 하나요 ? db에 저장되는 형태에 대한 정리가 필요할 것 같기도 합니다.. !! 타임 포맷을 정의하는 방식도 있을 수도 있을 것 같아요. ("yyyy-MM-dd'T'HH:mm:ss") 이런식으로요 ! 저도 이 부분에 대해서는 잘은 모르지만 요청이 이런 식으로 넘어가도 상관이 없는건지에 대한 의문이 듭니다.

yjy323 committed:
Merge branch 'develop' of https://github.com/Hae-il/BE into feat/consultation-api-#10

yjy323 committed:
Merge branch 'develop' of https://github.com/Hae-il/BE into feat/consultation-api-#10

yjy323 committed:
feat: 상담 기능 API 리팩토링 및 구조 개선 (#10)

- ConsultationRequest -> ConsultationReservation 도메인명 변경
- DTO 클래스명을 Request/Response 패턴으로 통일
- Client 엔티티를 별도 패키지로 분리
- 상담 예약 승인/거절 API 분리
- 상담 상태 관리 로직 단순화
- 파일 업로드 및 노트 관리 기능 개선
- 유효성 검증 어노테이션 추가

yjy323 commented on src/main/java/com/haeil/be/cases/service/CasesService.java:
이부분은 제꺼 먼저 develop에 merge 되면 준모님 코드랑 연계해서 수정할 것 같습니다.

yjy323 reviewed (commented):
상담 기능 API 수정 완료했습니다! ERD랑 API 명세서를 이어서 반영하겠습니다.
그리고 사건 API > 수임 API 순서대로 리뷰하면서 모순되는 부분 체크해둘게요!

Chominsung0823 reviewed (approved):
고생하셨습니다👍

yjy323 commented:
Swagger 테스트까지 확인했고 머지 하겠습니다~~
마참내 ㅜㅜ
